src/TokenBuyer.sol

```diff
@@ -0,0 +1,319 @@
+        }
+        _;
+    }
+
+    modifier onlyAdminOrOwner() {
+        if (admin != msg.sender && owner() != msg.sender) {
+            revert OnlyAdminOrOwner();
+        }
+        _;
+    }
+
+    constructor(
+        IERC20Metadata _paymentToken,
+        IOUToken _iouToken,
+        IPriceFeed _priceFeed,
+        uint256 _baselinePaymentTokenAmount,
+        uint256 _minAdminBaselinePaymentTokenAmount,
+        uint256 _maxAdminBaselinePaymentTokenAmount,
+        uint16 _botIncentiveBPs,
+        uint16 _minAdminBotIncentiveBPs,
+        uint16 _maxAdminBotIncentiveBPs,
+        address _owner,
+        address _admin,
+        address _payer
+    ) {
+        paymentToken = _paymentToken;
+        paymentTokenDecimalsDigits = 10**_paymentToken.decimals();
+        iouToken = _iouToken;
+        priceFeed = _priceFeed;
+
+        baselinePaymentTokenAmount = _baselinePaymentTokenAmount;
+        minAdminBaselinePaymentTokenAmount = _minAdminBaselinePaymentTokenAmount;
+        maxAdminBaselinePaymentTokenAmount = _maxAdminBaselinePaymentTokenAmount;
+
+        botIncentiveBPs = _botIncentiveBPs;
+        minAdminBotIncentiveBPs = _minAdminBotIncentiveBPs;
+        maxAdminBotIncentiveBPs = _maxAdminBotIncentiveBPs;
+
+        _transferOwnership(_owner);
+        admin = _admin;
+
+        payer = _payer;
+    }
+
+    /**
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      EXTERNAL TRANSACTIONS
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /**
+     * @notice Buy ETH from this contract in exchange for the ERC20 this token wants to acquire. The price
+     * is determined using `priceFeed` plus `botIncentiveBPs` basis points.
+     * @dev if `tokenAmount > tokenAmountNeeded()` uses the maximum amount possible. This function allows reentry because it does
+     * not allow double spending or exceeding the contract's {tokenAmountNeeded()}.
+     * @param tokenAmount the amount of ERC20 tokens msg.sender wishes to sell to this contract in exchange for ETH, in token decimals.
+     */
+    function buyETH(uint256 tokenAmount) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(tokenAmount, tokenAmountNeeded());
+
+        paymentToken.safeTransferFrom(msg.sender, payer, amount);
+
+        safeSendETH(msg.sender, ethAmountPerTokenAmount(amount), '');
+    }
+
+    function buyETH(
+        uint256 tokenAmount,
+        address to,
+        bytes calldata data
+    ) external nonReentrant whenNotPaused {
+        uint256 amount = Math.min(tokenAmount, tokenAmountNeeded());
+        uint256 balanceBefore = paymentToken.balanceOf(address(this));
+
+        safeSendETH(to, ethAmountPerTokenAmount(amount), abi.encode(msg.sender, amount, data));
+
+        uint256 tokensReceived = paymentToken.balanceOf(address(this)) - balanceBefore;
+        if (tokensReceived < amount) {
+            revert ReceivedInsufficientTokens(amount, tokensReceived);
+        }
+
+        paymentToken.safeTransfer(payer, tokensReceived);
```

Comment:
we could check that the `payer` already received tokens, instead of the current token, to avoid doing this transfer
wdyt?

Reply:
you're saying it's not awkward to expect buyers to pay a different contract (Payer) as they buy ETH from Buyer?

you might be right; my first stab at this felt like it makes sense to pay Buyer.

perhaps a 3rd solution: expect bots to approve Buyer to spend their tokens, then Buyer can stick to the same approach of running:
`paymentToken.safeTransferFrom(msg.sender, payer, amount);`

wdyt?

Reply:
the only point in doing my suggest is to reduce gas.

I think adding an approve action is again spending gas, so I don't know if it saves something

Reply:
From the searcher perspective, I'd rather send funds to the payer directly and avoid the extra transfer. Any security concerns with doing so?

Reply:
no security issue, just an awkwardness left over from before the refactor, and I felt like it might be better searcher UX to pay and get paid interacting with a single contract. 
your feedback that searchers will prefer the gas savings is great, will change

Reply:
fixed

Reply:
Only possible security concern I see is when there is a queued proposal to change the payer that's ready to be executed.

The proposal could be executed in the callback, so the before and after `balanceOf` calls check two different contracts. It will not be an issue unless the new payer has a balance and funds are migrated from the old payer in the proposal or the new payer has a greater balance than the existing payer.

Reply:
Solved by caching the payer as mentioned in David's comment.

Reply:
great thinking! fixed